### ISSUE TITLE
Support all payment types

### ISSUE THREAD

**Issue by qinghe** [sprangular/sprangular#39]
file module.coffee
`if !paymentMethods.gateway 
        alert 'Gateway is not configured in Spree...'`

why only check gateway?  what if check, COD?

**Comment by dapic** [comment]
In China, most online payments are done through the so called "offsite_payments", such as Alipay, Tenpay, Wxpay (WeChat payment)

**Comment by rposborne** [comment]
We are looking to support Paypal,   These clearly should be "extensions" that call in the relevant layout for the given payment gateway, if needed.    

I do think some code should be in the parent sprangular, namely methods for selecting gateways (tabs?) and a construct to pass relevant data to the extension if any processing or manipulation is required to support a given gateway request.

**Comment by rposborne** [comment]
@joshnuss what are you looking at for the timeline of the refactor to single page checkout?

**Comment by joshnuss** [comment]
@rpo The one page checkout refactoring will be ready early next week.

For PayPal Express, it can be done by adding an <a> tag that links to the paypal extensions checkout process (will direct user to paypal, and then back to store).  

Generally express checkout links (like PayPal, Google) would be better in the cart page, and bypass the current checkout, since PayPal can provide address data. (i.e. give user 2 checkout options "Checkout with PayPal", "Checkout with Credit Card"). 

We will need to add some extension points in the core to make this happen. PR welcome

**Comment by rposborne** [comment]
Agreed,  rightnow I am trying to make use of the existing better_paypal_express gem and trying to figure out a meaningful way to rewrite it's redirects into the JS application to make clean behavior vs straight reimplementation.  When I get a good handle I'll see where we can break up the implementation for flexibility and as a extension.

**Comment by joshnuss** [comment]
Cool.

Think the flow should work like this:

- On cart page, user clicks either "Checkout with Paypal" (or google checkout/amazon etc) or "Checkout with Credit Card" (use icons/images) (extract cart buttons to ng-include so its easy to override in new extensions)
- When user clicks "Checkout with PayPal", they are redirected to paypal's checkout screen
- Once user enters details on PayPal, they click "Pay now" and are redirected back to the sprangular app's "express checkout" page.
- User selects shipping rate, and confirms order details (tax, summary etc)
- User click "Place Order", the shipping rate and tax rate are sent over to paypal and money is debited from PayPal account
- User is redirected to /checkout/complete page

**Comment by rposborne** [comment]
We are tracking a paypal extension here. 
https://github.com/feldfix/sprangular_paypal_express/tree/2-4-stable

There are currently some issues but primary communication is present.

**Comment by joshnuss** [comment]
Groovy :+1:
